ext/ffi/ffi.c

```diff
@@ -4382,54 +4382,73 @@ ZEND_METHOD(FFI, alignof) /* {{{ */
 }
 /* }}} */
 
+static void *php_ffi_parse_cdata_to_ptr(zend_long size, zend_object *zobj, bool *is_error)
+{
+	void *result_ptr = NULL;
+	*is_error = false;
+
+	zend_ffi_cdata *cdata = (zend_ffi_cdata*)zobj;
+    zend_ffi_type *type = ZEND_FFI_TYPE(cdata->type);
+
+    if (type->kind == ZEND_FFI_TYPE_POINTER) {
+    	result_ptr = *(void**)cdata->ptr;
+    } else {
+    	if (type->kind != ZEND_FFI_TYPE_POINTER && size > type->size) {
+    		*is_error = true;
+    		return NULL;
+    	}
+    	result_ptr = cdata->ptr;
+    }
+
+    return result_ptr;
+}
+
+static void *php_ffi_parse_cdata_or_string_to_ptr(zend_long size, zend_object *zobj, zend_string *zstr, bool *is_error)
+{
+	void *result_ptr = NULL;
+	*is_error = false;
+
+	if (zstr) {
+		if (size > ZSTR_LEN(zstr)) {
+    		*is_error = true;
+			return NULL;
+		}
+		result_ptr = ZSTR_VAL(zstr);
+	} else {
+		result_ptr = php_ffi_parse_cdata_to_ptr(size, zobj, is_error);
+	}
+
+    return result_ptr;
+}
+
```

Comment:
The separation of this functions doesn't make things more clear, because you delegated them to check for data/string size and this became unclear.